### PR TITLE
feat(dsx): Introduce EMQX durable storage explorer

### PR DESCRIPTION
**Pull request by ieQu1**
Fixes <issue-or-jira-number>

Release version: v/e5.?

## PR Checklist
Please convert it to a draft if any of the following conditions are not met. Reviewers may skip over until all the items are checked:

- [ ] Added tests for the changes
- [ ] Added property-based tests for code which performs user input validation
- [ ] Changed lines covered in coverage report
- [ ] Change log has been added to `changes/(ce|ee)/(feat|perf|fix|breaking)-<PR-id>.en.md` files
- [ ] For internal contributor: there is a jira ticket to track this change
- [ ] Created PR to [emqx-docs](https://github.com/emqx/emqx-docs) if documentation update is required, or link to a follow-up jira ticket
- [ ] Schema changes are backward compatible

## Checklist for CI (.github/workflows) changes

- [ ] If changed package build workflow, pass [this action](https://github.com/emqx/emqx/actions/workflows/build_packages.yaml) (manual trigger)
- [ ] Change log has been added to `changes/` dir for user-facing artifacts update

### CODE REVIEW

**Review by savonarola — APPROVED** [review]

File: apps/emqx_durable_storage/src/emqx_dsx.erl [hunk 1]
```diff
@@ -0,0 +1,380 @@
...
+    renew_streams_interval :: pos_integer(),
+    %%   Poll-related fields:
+    %% In-progress poll requests:
+    polls = #{} :: #{emqx_ds:stream() => reference()},
+    %% Queue of pollable streams:
+    pq :: queue:queue(emqx_ds:stream()),
+    %% Number of streams with currently inflight poll requests:
+    inflight = 0 :: non_neg_integer(),
+    %% Number of poll requests sent to the DS so far, or since the
+    %% last `next{}' call:
```
Comment (thalesmg) [inline apps/emqx_durable_storage/src/emqx_dsx.erl:200]:
```suggestion
    %% last `#more_req{}' call:
```

File: apps/emqx_durable_storage/src/emqx_dsx.erl [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+    S = #s{
+        callback = Callback,
+        db = DB,
+        topic = TF,
+        start = Start,
+        active = Active,
+        renew_streams_interval = RSI,
+        poll_timeout = PollTimeout,
+        pq = queue:new()
+    },
```
Comment (thalesmg) [inline apps/emqx_durable_storage/src/emqx_dsx.erl:225]: It seems like `callback_state` is never initialized?

File: apps/emqx_durable_storage/src/emqx_dsx.erl [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+        callback_state = CS
+    };
+handle_poll_reply(Stream, {error, recoverable, _Err}, S = #s{pq = PQ0}) ->
+    S#s{pq = queue:in(Stream, PQ0)};
+handle_poll_reply(Stream, EOSEvent, S = #s{eos = EOS}) ->
+    case EOSEvent of
+        {ok, end_of_stream} ->
+            ok;
+        {error, unrecoverable, _Err} ->
+            ok
```
Comment (thalesmg) [inline apps/emqx_durable_storage/src/emqx_dsx.erl:307]: Maybe log the error?

**Review by qzhuyan — DISMISSED** [review]

**Review by thalesmg — DISMISSED** [review]

File: apps/emqx_durable_storage/src/emqx_dsx.erl [hunk 4]
```diff
@@ -0,0 +1,380 @@
...
+    User = maps:update_with(
+        topic,
+        fun(Topic) ->
+            case Topic of
+                _ when is_binary(Topic) ->
+                    emqx_topic:words(Topic);
+                [A | _] when is_atom(A); is_binary(A) ->
+                    Topic;
+                [A | _] when is_integer(A) ->
+                    emqx_topic:words(list_to_binary(Topic))
```
[the remark this answers is not in the document]
Comment (ieQu1) [reply apps/emqx_durable_storage/src/emqx_dsx.erl:128]: Good catch, typespec is in fact wrong.

File: apps/emqx_durable_storage/src/emqx_dsx.erl (hunk 2 above)
Comment (ieQu1) [reply apps/emqx_durable_storage/src/emqx_dsx.erl:225]: Initially it's undefined.

File: apps/emqx_durable_storage/src/emqx_dsx.erl (hunk 3 above)
Comment (ieQu1) [reply apps/emqx_durable_storage/src/emqx_dsx.erl:307]: Done

File: apps/emqx_durable_storage/src/emqx_dsx.erl [hunk 5]
```diff
@@ -0,0 +1,384 @@
...
+handle_info(
+    #poll_reply{ref = Ref, userdata = Stream, payload = Payload},
+    #s{inflight = Inflight, polls = Polls0} = S0
+) ->
+    case maps:take(Stream, Polls0) of
+        {Ref, Polls} ->
+            S = handle_poll_reply(Stream, Payload, S0#s{inflight = Inflight - 1, polls = Polls}),
+            {noreply, poll(S)};
+        _ ->
+            logger:error("Stray poll reply ~p", [
```
Comment (savonarola) [inline apps/emqx_durable_storage/src/emqx_dsx.erl:264]: Is it intentional not to use SLOG and snake messages for logs?

Comment (ieQu1) [reply apps/emqx_durable_storage/src/emqx_dsx.erl:264]: Yes. SLOG is declared in emqx application, this code is part of emqx_durable_storage. DS cannot depend on EMQX code.

File: apps/emqx_durable_storage/src/emqx_dsx.erl [hunk 6]
```diff
@@ -0,0 +1,384 @@
...
+
+update_stats(Worker, {_Key, Msg}) ->
+    emqx_metrics_worker:inc(Worker, Worker, n_msgs),
+    emqx_metrics_worker:inc(Worker, Worker, n_bytes, emqx_message:estimate_size(Msg)),
+    Lag = emqx_message:timestamp_now() - Msg#message.timestamp,
+    emqx_metrics_worker:observe(Worker, Worker, lag, Lag).
+
+%%%%%
+-spec start_link(#{
+    topic := string() | [emqx_types:word()],
```
Comment (thalesmg) [inline apps/emqx_durable_storage/src/emqx_dsx.erl:110]:
```suggestion
    topic := binary() | string() | [emqx_types:word()],
```

**Review by thalesmg — APPROVED** [review]

File: apps/emqx_durable_storage/src/emqx_dsx.erl (hunk 2 above)
Comment (ieQu1) [reply apps/emqx_durable_storage/src/emqx_dsx.erl:226]:
```
case State of 
  undefined -> init()
  _ -> ...
end
```

If this is confusing, I can remove the state completely.

**Pull request merged**